src/resolvers/Promotion.js

```diff
@@ -0,0 +1,267 @@
+       * - check validity of promotion name
+       * - disallow name if the promotion is still ongoing
+       */
+      const existingPromotion = await Promotion.findOne({
+        promotionNm: promotionInput.promotionNm,
+      });
+      if (
+        existingPromotion &&
+        new Date(existingPromotion.promotionEnd) > Date.now()
+      ) {
+        throw new ApolloError("Prmotion name alread taken");
+      }
+
+      /**
+       * date range validity check:
+       * - must start at least one week from now
+       * - end date > start date
+       */
+      const validStartDt =
+        new Date(promotionInput.promotionStart) >
+        new Date(Date.now() + 1000 * 60 * 60 * 24 * 7);
+      const validEndDt =
+        new Date(promotionInput.promotionStart) <
+        new Date(promotionInput.promotionEnd);
+      if (!validStartDt || !validEndDt) {
+        throw UserInputError("Invalid date");
+      }
+
+      promotionInput.image._id = promotionInput.image.id;
+      promotionInput.promotionService.forEach((element) => {
+        element._id = element.id;
+      });
+      promotionInput.promotionBranch.forEach((element) => {
+        element._id = element.id;
+      });
+
+      let newPromotion = await Promotion.create(promotionInput);
+
+      newPromotion = promotionConverter(newPromotion);
+      newPromotion.id = newPromotion._id;
+
+      return newPromotion;
+    },
+    updatePromotion: async (root, { promotionInput }, context, info) => {
+      /**
+       * - only allow loggedin brandowner
+       * - valid id check
+       * - name validity check
+       * - date validity check
+       */
+
+      /**
+       * check user type is Brand owner
+       */
+      if (context.user.type !== USER_TYPE.BRANDOWNER) {
+        throw AuthenticationError("Invalid Permission");
+      }
+
+      const validId = mongoose.Types.ObjectId.isValid(promotionInput.id);
+      const existingPromotionById = await Promotion.findById(promotionInput.id);
+      if (!validId || existingPromotionById === null) {
+        throw new UserInputError("Invalid ID");
+      }
+
+      /**
+       * - check validity of promotion name
+       * - disallow name if the promotion is still ongoing
+       * - allow if the name is same as previous name (means no update)
+       */
+      const existingPromotion = await Promotion.findOne({
+        promotionNm: { $eq: promotionInput.promotionNm },
+      });
+      if (
+        existingPromotion &&
+        existingPromotion.promotionNm !== promotionInput.promotionNm
+      ) {
+        if (new Date(existingPromotion.promotionEnd) > Date.now()) {
+          throw new ApolloError("Prmotion name alread taken");
+        }
+      }
+
+      /**
+       * date range validity check:
+       * - must start at least one week from now
+       * - end date > start date
+       */
+      const validStartDt =
+        new Date(promotionInput.promotionStart) >
+        new Date(Date.now() + 1000 * 60 * 60 * 24 * 7);
```

Comment:
when i update promotion but i did not update the start date?
example scenario:
i have I promotion with startDate (27/01/2021) that created last week(15/01/2021)
i want to update my promotion today (22/1/2021) to rename the name maybe
what will happen? because my promotion will still need to start at 27/01/2021

Reply:
you are right. maybe update should not check date validity

Reply:
Maybe we check if the 
- **new start  date**  > 7 days from now
OR
- **new start date** later or equal **original start date**